apis/k8ssandra/v1alpha1/k8ssandracluster_webhook_test.go

```diff
@@ -319,6 +320,114 @@ func testNumTokens(t *testing.T) {
+	newCluster = createCluster(nil, []string{"rack1", "rack2"})
+	required.NoError(newCluster.ValidateUpdate(oldCluster))
+
+	oldCluster = createCluster(nil, nil)
+	newCluster = createCluster(nil, []string{"default"})
+	required.NoError(newCluster.ValidateUpdate(oldCluster))
+
+	// Same scenarios, but top-level declaration
+	oldCluster = createCluster([]string{"rack1", "rack2"}, nil)
+	newCluster = createCluster([]string{"rack1", "rack2New"}, nil)
+	required.EqualError(newCluster.ValidateUpdate(oldCluster),
+		"racks can't be renamed (DC dc1, index 1, current=rack2, desired=rack2New)")
+
+	oldCluster = createCluster([]string{"rack1", "rack2"}, nil)
+	newCluster = createCluster([]string{"rack1"}, nil)
+	required.EqualError(newCluster.ValidateUpdate(oldCluster),
+		"number of racks can't be lowered (DC dc1, current=2, desired=1)")
+
+	oldCluster = createCluster(nil, nil)
+	newCluster = createCluster([]string{"rack1"}, nil)
+	required.EqualError(newCluster.ValidateUpdate(oldCluster),
+		"racks can't be renamed (DC dc1, index 0, current=default, desired=rack1)")
+
+	oldCluster = createCluster([]string{"rack1"}, nil)
+	newCluster = createCluster(nil, nil)
+	required.EqualError(newCluster.ValidateUpdate(oldCluster),
+		"racks can't be renamed (DC dc1, index 0, current=rack1, desired=default)")
+
+	oldCluster = createCluster([]string{"rack1"}, nil)
+	newCluster = createCluster([]string{"rack1", "rack2"}, nil)
+	required.NoError(newCluster.ValidateUpdate(oldCluster))
+
+	oldCluster = createCluster(nil, nil)
+	newCluster = createCluster([]string{"default"}, nil)
+	required.NoError(newCluster.ValidateUpdate(oldCluster))
+
+	// Mix of top-level/DC-level, error scenarios
+	oldCluster = createCluster([]string{"rack1", "rack2"}, nil)
+	newCluster = createCluster(nil, []string{"rack1", "rack2New"})
+	required.EqualError(newCluster.ValidateUpdate(oldCluster),
+		"racks can't be renamed (DC dc1, index 1, current=rack2, desired=rack2New)")
+
+	oldCluster = createCluster(nil, []string{"rack1", "rack2"})
+	newCluster = createCluster([]string{"rack1", "rack2New"}, nil)
+	required.EqualError(newCluster.ValidateUpdate(oldCluster),
+		"racks can't be renamed (DC dc1, index 1, current=rack2, desired=rack2New)")
+
+	// Mix of top-level/DC-level, valid scenarios
+	oldCluster = createCluster([]string{"rack1"}, nil)
+	newCluster = createCluster(nil, []string{"rack1"})
+	required.NoError(newCluster.ValidateUpdate(oldCluster))
+
+	oldCluster = createCluster(nil, []string{"rack1"})
+	newCluster = createCluster([]string{"rack1"}, nil)
+	required.NoError(newCluster.ValidateUpdate(oldCluster))
+
+	oldCluster = createCluster([]string{"rack1"}, nil)
+	newCluster = createCluster(nil, []string{"rack1", "rack2"})
+	required.NoError(newCluster.ValidateUpdate(oldCluster))
+
+	oldCluster = createCluster(nil, []string{"rack1"})
+	newCluster = createCluster([]string{"rack1", "rack2"}, nil)
+	required.NoError(newCluster.ValidateUpdate(oldCluster))
```

Comment:
From these tests, it isn't quite clear to me what should happen when you have multiple DCs. 

I'd almost like to suggest that you take a step back and stop providing the DC and cluster level racks. Instead perhaps you could provide the whole K8ssandraCluster struct to make sure that the coalescing logic is all covered?